ryhope/src/storage/memory.rs

```diff
@@ -286,53 +357,218 @@ where
+                assert_eq!(last_epoch, matched_epoch,);
+                None
+            }
+            None => {
+                // get arbitrary epoch corresponding to the new incremental epoch.
+                // in this implementation, it is computed assuming that also
+                // `UserEpoch`s are incremental, and so the epoch to be inserted
+                // is simply `last_epoch + 1`
+                let mapped_epoch = last_epoch + 1;
+                // add the epoch mapping to `self`
+                self.add_epoch(mapped_epoch, epoch)
+                    .ok()
+                    .map(|_| mapped_epoch)
+            }
+        }
+    }
+
+    pub(crate) fn rollback_to(&mut self, epoch: UserEpoch) {
+        // erase from the map all epochs greater than `epoch`
+        let to_be_erased_epochs = self
+            .0
+            .iter()
+            .rev()
+            .map_while(|el| if *el.0 > epoch { Some(*el.0) } else { None })
+            .collect_vec();
+        to_be_erased_epochs.into_iter().for_each(|epoch| {
+            self.0.remove(&epoch);
+        });
+    }
+
+    fn add_epoch(
+        &mut self,
+        user_epoch: UserEpoch,
+        incremental_epoch: IncrementalEpoch,
+    ) -> Result<()> {
+        // double check that we are either replacing an existing `IncrementalEpoch`
+        // in the map or we are adding the next incremental one. This check ensures
+        // that `IncrementalEpoch`s found in the map are always incremental
+        let num_epochs = self.0.len();
         ensure!(
-            epoch <= self.current_epoch(),
-            "unable to rollback to epoch `{}` more recent than current epoch `{}`",
-            epoch,
-            self.current_epoch()
+            incremental_epoch as usize <= num_epochs,
+            "Inserted IncrementalEpoch is too big: found {incremental_epoch}, maximum is {num_epochs}"
         );
+        // check that the `user_epoch` being added is associated to the correct
+        // `incremental_epoch`, according to the ordering in the map
+        if let Some(smaller_epoch) = self.try_to_user_epoch_inner(incremental_epoch - 1) {
+            ensure!(user_epoch > smaller_epoch);
+        }
+        if let Some(bigger_epoch) = self.try_to_user_epoch_inner(incremental_epoch + 1) {
+            ensure!(user_epoch < bigger_epoch)
+        }
+        // if we are replacing an existing `IncrementalEpoch`, ensure that
+        // we remove the old mapping entry
+        if let Some(epoch) = self.try_to_user_epoch_inner(incremental_epoch) {
+            self.0.remove(&epoch);
+        }
 
-        self.mem.truncate((epoch + 1).try_into().unwrap());
-
+        self.0.insert(user_epoch, incremental_epoch);
         Ok(())
     }
 }
 
+impl EpochMapper for InMemoryEpochMapper {
+    async fn try_to_incremental_epoch(&self, epoch: UserEpoch) -> Option<IncrementalEpoch> {
+        self.try_to_incremental_epoch_inner(epoch)
```

Comment:
Could we put these function bodies right here, or is there a reason to implement them in duplicate out of the trait implementation?

Reply:
You are right, I think I initially had an inner non-async method to be able to call this function also in other non-async methods, but then I had to made also those methods async for other reasons and so I resorted to call directly the async method of the `EpochMapper` trait, but I forgot to remove the inner non-async implementation.